// may-youtubei/dist/esm/youtube/BaseVideo/BaseVideoParser.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { getContinuationFromItems, stripToInt, Thumbnails } from "../../common";
import { BaseChannel } from "../BaseChannel";
import { PlaylistCompact } from "../PlaylistCompact";
import { VideoCompact } from "../VideoCompact";
var BaseVideoParser = /** @class */ (function () {
    function BaseVideoParser() {
    }
    BaseVideoParser.loadBaseVideo = function (target, data) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
        var videoInfo = BaseVideoParser.parseRawData(data);
        // Basic information
        target.id = videoInfo.videoDetails.videoId;
        target.title = videoInfo.videoDetails.title;
        target.uploadDate = videoInfo.dateText.simpleText;
        target.uploadDateNoTz = videoInfo.microformat.playerMicroformatRenderer.uploadDate;
        target.viewCount = +videoInfo.videoDetails.viewCount || null;
        target.isLiveContent = videoInfo.videoDetails.isLiveContent;
        target.thumbnails = new Thumbnails().load(videoInfo.videoDetails.thumbnail.thumbnails);
        var _m = ((_g = (_f = (_e = (_d = (_c = (_b = (_a = videoInfo.metadataRowContainer) === null || _a === void 0 ? void 0 : _a.metadataRowContainerRenderer) === null || _b === void 0 ? void 0 : _b.rows) === null || _c === void 0 ? void 0 : _c.at(0)) === null || _d === void 0 ? void 0 : _d.richMetadataRowRenderer) === null || _e === void 0 ? void 0 : _e.contents) === null || _f === void 0 ? void 0 : _f.at(0)) === null || _g === void 0 ? void 0 : _g.richMetadataRenderer) || {}, endpoint = _m.endpoint, gameTitle = _m.title;
        target.gameChannelId = ((_h = endpoint === null || endpoint === void 0 ? void 0 : endpoint.browseEndpoint) === null || _h === void 0 ? void 0 : _h.browseId) || null;
        target.gameName = (gameTitle === null || gameTitle === void 0 ? void 0 : gameTitle.simpleText) || null;
        // Channel
        var _o = videoInfo.owner.videoOwnerRenderer, title = _o.title, thumbnail = _o.thumbnail, subscriberCountText = _o.subscriberCountText;
        target.channel = new BaseChannel({
            client: target.client,
            id: title.runs[0].navigationEndpoint.browseEndpoint.browseId,
            name: title.runs[0].text,
            subscriberCount: subscriberCountText === null || subscriberCountText === void 0 ? void 0 : subscriberCountText.simpleText,
            thumbnails: new Thumbnails().load(thumbnail.thumbnails),
        });
        // Like Count and Dislike Count
        var topLevelButtons = videoInfo.videoActions.menuRenderer.topLevelButtons;
        target.likeCount = stripToInt(BaseVideoParser.parseButtonRenderer(topLevelButtons[0]));
        // Tags and description
        target.tags =
            ((_k = (_j = videoInfo.superTitleLink) === null || _j === void 0 ? void 0 : _j.runs) === null || _k === void 0 ? void 0 : _k.map(function (r) { return r.text.trim(); }).filter(function (t) { return t; })) || [];
        target.description =
            ((_l = videoInfo.description) === null || _l === void 0 ? void 0 : _l.runs.map(function (d) { return d.text; }).join("")) || "";
        // related videos
        var secondaryContents = data[3].response.contents.twoColumnWatchNextResults.secondaryResults.secondaryResults
            .results;
        if (secondaryContents) {
            target.related.items = BaseVideoParser.parseRelatedFromSecondaryContent(secondaryContents, target.client);
            target.related.continuation = getContinuationFromItems(secondaryContents);
        }
        return target;
    };
    BaseVideoParser.parseRelated = function (data, client) {
        var secondaryContents = data.onResponseReceivedEndpoints[0].appendContinuationItemsAction.continuationItems;
        return BaseVideoParser.parseRelatedFromSecondaryContent(secondaryContents, client);
    };
    BaseVideoParser.parseContinuation = function (data) {
        var secondaryContents = data.onResponseReceivedEndpoints[0].appendContinuationItemsAction.continuationItems;
        return getContinuationFromItems(secondaryContents);
    };
    BaseVideoParser.parseRawData = function (data) {
        var contents = data[3].response.contents.twoColumnWatchNextResults.results.results.contents;
        var primaryInfo = contents.find(function (c) { return "videoPrimaryInfoRenderer" in c; })
            .videoPrimaryInfoRenderer;
        var secondaryInfo = contents.find(function (c) { return "videoSecondaryInfoRenderer" in c; }).videoSecondaryInfoRenderer;
        var _a = data[2].playerResponse, videoDetails = _a.videoDetails, microformat = _a.microformat;
        return __assign(__assign(__assign({}, secondaryInfo), primaryInfo), { videoDetails: videoDetails, microformat: microformat });
    };
    BaseVideoParser.parseCompactRenderer = function (data, client) {
        if ("compactVideoRenderer" in data) {
            return new VideoCompact({ client: client }).load(data.compactVideoRenderer);
        }
        else if ("compactRadioRenderer" in data) {
            return new PlaylistCompact({ client: client }).load(data.compactRadioRenderer);
        }
    };
    BaseVideoParser.parseRelatedFromSecondaryContent = function (secondaryContents, client) {
        return secondaryContents
            .map(function (c) { return BaseVideoParser.parseCompactRenderer(c, client); })
            .filter(function (c) { return c !== undefined; });
    };
    BaseVideoParser.parseButtonRenderer = function (data) {
        var _a;
        var buttonRenderer;
        if (!data.segmentedLikeDislikeButtonRenderer) {
            buttonRenderer = data.toggleButtonRenderer || data.buttonRenderer;
        }
        else {
            var likeButton = data.segmentedLikeDislikeButtonRenderer.likeButton;
            buttonRenderer = likeButton.toggleButtonRenderer || likeButton.buttonRenderer;
        }
        var accessibilityData = (((_a = buttonRenderer.defaultText) === null || _a === void 0 ? void 0 : _a.accessibility) || buttonRenderer.accessibilityData).accessibilityData;
        return accessibilityData.label;
    };
    return BaseVideoParser;
}());
export { BaseVideoParser };
