// @fabricio-191/youtube/lib/methods/video.js
const { getID, parseOptions, requests } = require('../utils/utils.js');
const { parse, Utils, parsers } = require('../parser/main.js');
// const parseStreamingData = require('../download/formats.js');

async function getVideo(URLorID, options){
	const parsedOptions = parseOptions(options);
	const body = await requests.fetch(
		`https://www.youtube.com/watch?v=${getID(URLorID, 1)}`,
		parsedOptions
	// @ts-ignore
	).text();
	const data = requests.getData(body, 1), playerResponse = requests.getData(body, 2);

	if(!playerResponse.videoDetails) return null;

	try{
		return videoInfo(data, playerResponse);
	}catch(e){
		Object.assign(e, {
			playabilityStatus: playerResponse.playabilityStatus.status,
			videoURL: `https://www.youtube.com/watch?v=${playerResponse.videoDetails.videoId}`,
			loggedIn: !playerResponse.responseContext.mainAppWebResponseContext.loggedOut,
			options: parsedOptions,
		});

		throw e;
	}
}

module.exports = getVideo;

function videoInfo(data, playerResponse){
	data = data.contents.twoColumnWatchNextResults;
	const mainResults = data.results.results.contents;
	const { videoPrimaryInfoRenderer } = mainResults.find(v => v.videoPrimaryInfoRenderer);
	const { videoSecondaryInfoRenderer } = mainResults.find(v => v.videoSecondaryInfoRenderer);

	const info = {
		ID: playerResponse.videoDetails.videoId,
		URL: `https://www.youtube.com/watch?v=${playerResponse.videoDetails.videoId}`,
		name: Utils.parseText(videoPrimaryInfoRenderer.title).toString(),

		likes: getLikes(videoPrimaryInfoRenderer),

		views: new Utils.Views(videoPrimaryInfoRenderer.viewCount),
		owner: parse(videoSecondaryInfoRenderer.owner),

		description: Utils.parseText(videoSecondaryInfoRenderer.description),
		thumbnails: new Utils.Thumbnails(playerResponse.videoDetails.thumbnail),
		keywords: playerResponse.videoDetails.keywords || null,
		uploadDateLabel: Utils.parseText(videoPrimaryInfoRenderer.dateText).toString(),
		isLive: playerResponse.videoDetails.isLiveContent,
		...parse(playerResponse.microformat),
		duration: new Utils.Duration(playerResponse.videoDetails.lengthSeconds),
	};

	const endScreen = Utils.optionalChaining(data, 'playerOverlays.playerOverlayRenderer.endScren.watchNextEndScreenRenderer');
	if(endScreen && endScreen.results) info.endScreen = {
		title: Utils.parseText(endScreen.title).toString(),
		items: endScreen.results.map(parse),
	};

	const secondaryResults = Utils.optionalChaining(data, 'secondaryResults.secondaryResults.results');
	if(secondaryResults && secondaryResults.length > 1){
		if(secondaryResults[secondaryResults.length -1].continuationItemRenderer){
			secondaryResults.pop();
		}

		info.related = secondaryResults.filter(a => !a.compactAutoplayRenderer).map(parse);
	}

	if(data.playlist) info.playlist = {
		ID: data.playlist.playlistId,
		title: Utils.parseText(data.playlist.titleText).toString(),
		owner: parsers.bylineText(data.playlist),
		videoQuantity: data.playlist.totalVideos,
		videos: data.playlist.videos.map(playlistVideo),
	};

	if(playerResponse.streamingData){
		// Object.assign(info, parseStreamingData(playerResponse.streamingData));
	}

	return info;
}

function playlistVideo({ playlistPanelVideoRenderer }){
	return {
		name: Utils.parseText(playlistPanelVideoRenderer.title).toString(),
		ID: playlistPanelVideoRenderer.videoId,
		playlistID: playlistPanelVideoRenderer.navigationEndpoint.watchEndpoint.playlistId,
		URL: `https://www.youtube.com${playlistPanelVideoRenderer.navigationEndpoint.commandMetadata.webCommandMetadata.url}`,

		duration: new Utils.Duration(playlistPanelVideoRenderer),
		views: new Utils.Views(playlistPanelVideoRenderer),
		thumbnails: new Utils.Thumbnails(playlistPanelVideoRenderer.thumbnail),

		publishedDate: Utils.parseText(playlistPanelVideoRenderer.publishedTimeText).toString(),

		owner: parsers.bylineText(playlistPanelVideoRenderer),
	};
}

function getLikes(videoPrimaryInfoRenderer){
	let likeButton = Utils.optionalChaining(videoPrimaryInfoRenderer, 'videoActions.menuRenderer.topLevelButtons.0');
	if(likeButton.segmentedLikeDislikeButtonRenderer){
		// eslint-disable-next-line prefer-destructuring
		likeButton = likeButton.segmentedLikeDislikeButtonRenderer.likeButton;
	}

	return Utils.extractInt(likeButton.toggleButtonRenderer.defaultText) || 0;
}
